src/hotspot/cpu/s390/c1_LIRAssembler_s390.cpp

```diff
@@ -1532,8 +1532,20 @@ void LIR_Assembler::arith_op(LIR_Code code, LIR_Opr left, LIR_Opr right, LIR_Opr
       // cpu register - constant
       jint c = right->as_constant_ptr()->as_jint();
       switch (code) {
-        case lir_add: __ z_agfi(lreg, c);  break;
-        case lir_sub: __ z_agfi(lreg, -c); break; // note: -min_jint == min_jint
+        case lir_add:
+                      if (Immediate::is_simm16(c)) {
+                        __ z_ahi(lreg, c);
+                      } else {
+                        __ z_afi(lreg, c);
+                      }
```

Comment:
It seems like it would be better to have code like this in a helper function, instead of making every call site repeat the pattern.  Can you use add2reg() here?

Reply:
add2reg will emit `z_aghi` instruction which is for 64 bit (register) <- 16 bit (immediate). `z_ahi` and `z_afi` are both 
32 bit (register) < - 16 bit (immediate), 32 bit (register) <- 32 bit (immediate). So I guess these are better here. Though we can move the logic to new method add2reg_32() method which will do the check and emit correct instruction.